File: care/emr/reports/context_builder/data_points/symptom.py
from django_filters import rest_framework as filters

from care.emr.models.condition import Condition
from care.emr.reports.context_builder.data_points.base import (
    Field,
    QuerysetContextBuilder,
)
from care.emr.reports.context_builder.data_points.user import SingleUserIdContextBuilder
from care.emr.resources.condition.spec import CategoryChoices
from care.utils.filters.multiselect import MultiSelectFilter

CLINICAL_STATUS_DISPLAY = {
    "active": "Active",
    "recurrence": "Recurrence",
    "relapse": "Relapse",
    "inactive": "Inactive",
    "remission": "Remission",
    "resolved": "Resolved",
    "unknown": "Unknown",
}

VERIFICATION_STATUS_DISPLAY = {
    "unconfirmed": "Unconfirmed",
    "provisional": "Provisional",
    "differential": "Differential",
    "confirmed": "Confirmed",
    "refuted": "Refuted",
    "entered_in_error": "Entered in Error",
}


class SymptomsReportFilter(filters.FilterSet):
    clinical_status = MultiSelectFilter(field_name="clinical_status")
    exclude_clinical_status = MultiSelectFilter(
        field_name="clinical_status", exclude=True
    )
    verification_status = MultiSelectFilter(field_name="verification_status")
    exclude_verification_status = MultiSelectFilter(
        field_name="verification_status", exclude=True
    )


class SymptomsContextBuilder(QuerysetContextBuilder):
    filterset_class = SymptomsReportFilter
    __filterset_backends__ = [filters.DjangoFilterBackend]

    clinical_status = Field(
        display="Clinical Status",
        preview_value="Active",
        mapping=lambda c: CLINICAL_STATUS_DISPLAY.get(
            c.clinical_status, c.clinical_status.title()
        )
        if c.clinical_status
        else "",
        description="Clinical status of the condition",
    )
    verification_status = Field(
        display="Verification Status",
        preview_value="Confirmed",
        mapping=lambda c: VERIFICATION_STATUS_DISPLAY.get(
            c.verification_status, c.verification_status.title()
        )
        if c.verification_status
        else "",
        description="Verification status of the condition",
    )
    name = Field(
        display="Name",
        mapping=lambda c: c.code.get("display") if c.code else "",
        preview_value="Fever",
        description="Name of the symptom",
    )

    onset = Field(
        display="Onset",
        mapping=lambda c: c.onset.get("onset_datetime") if c.onset else "",
        preview_value="2025-11-30T18:30:00Z",
        description="The onset date of the symptom",
    )

    note = Field(
        display="Note",
        preview_value="",
        description="Additional notes about the symptom",
    )
    created_by = Field(
        display="Created By",
        target_context=SingleUserIdContextBuilder,
        preview_value="",
        description="User who created the symptom",
    )
    updated_by = Field(
        display="Updated By",
        target_context=SingleUserIdContextBuilder,
        preview_value="",
        description="User who updated the symptom",
    )

    def get_context(self):
        return Condition.objects.filter(
            encounter=self.parent_context,
            category=CategoryChoices.problem_list_item.value,
        )

File: frontend/src/types/emr/symptom/symptom.ts
import { Code } from "@/types/base/code/code";
import { UserReadMinimal } from "@/types/user/user";

export const SYMPTOM_CLINICAL_STATUS = [
  "active",
  "recurrence",
  "relapse",
  "inactive",
  "remission",
  "resolved",
] as const;

export type SymptomClinicalStatus = (typeof SYMPTOM_CLINICAL_STATUS)[number];

export const SYMPTOM_VERIFICATION_STATUS = [
  "unconfirmed",
  "provisional",
  "differential",
  "confirmed",
  "refuted",
  "entered_in_error",
] as const;

export type SymptomVerificationStatus =
  (typeof SYMPTOM_VERIFICATION_STATUS)[number];

export const SYMPTOM_SEVERITY = ["severe", "moderate", "mild"] as const;

export type SymptomSeverity = (typeof SYMPTOM_SEVERITY)[number];

export type Onset = {
  onset_datetime?: string;
  onset_age?: string;
  onset_string?: string;
  note?: string;
};

export interface Symptom {
  id: string;
  code: Code;
  clinical_status: SymptomClinicalStatus;
  verification_status: SymptomVerificationStatus;
  severity: SymptomSeverity;
  onset?: Onset;
  recorded_date?: string;
  note?: string;
  created_by: UserReadMinimal;
  updated_by: UserReadMinimal;
  category: string;
  encounter: string;
  created_date: string;
  updated_date?: string;
}

export interface SymptomRequest {
  id?: string;
  clinical_status: SymptomClinicalStatus;
  verification_status: SymptomVerificationStatus;
  code: Code;
  severity: SymptomSeverity;
  onset?: Onset;
  recorded_date?: string;
  note?: string;
  encounter: string;
  category: string;
  created_date?: string;
  updated_date?: string;
  created_by?: UserReadMinimal;
}

export const SYMPTOM_CLINICAL_STATUS_COLORS = {
  active: "primary",
  recurrence: "yellow",
  relapse: "destructive",
  inactive: "secondary",
  remission: "blue",
  resolved: "green",
} as const;

export const SYMPTOM_VERIFICATION_STATUS_COLORS = {
  unconfirmed: "yellow",
  provisional: "orange",
  differential: "purple",
  confirmed: "green",
  refuted: "destructive",
  entered_in_error: "destructive",
} as const;

export const SYMPTOM_SEVERITY_COLORS = {
  severe: "destructive",
  moderate: "yellow",
  mild: "blue",
} as const;
